runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkExecutableStageFunction.java

```diff
@@ -166,40 +180,126 @@ private StateRequestHandler getStateRequestHandler(
+                windowCoder));
+
+    try (RemoteBundle bundle =
+        stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler)) {
+      processElements(iterable, bundle);
+    }
+
+    // Finish any pending windows by advancing the input watermark to infinity.
+    timerInternals.advanceInputWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE);
+    // Finally, advance the processing time to infinity to fire any timers.
+    timerInternals.advanceProcessingTime(BoundedWindow.TIMESTAMP_MAX_VALUE);
+    timerInternals.advanceSynchronizedProcessingTime(BoundedWindow.TIMESTAMP_MAX_VALUE);
+
+    try (RemoteBundle bundle =
+        stageBundleFactory.getBundle(receiverFactory, stateRequestHandler, progressHandler)) {
+
+      fireEligibleTimers(
+          timerInternals,
+          (String timerId, WindowedValue timerValue) -> {
+            FnDataReceiver<WindowedValue<?>> fnTimerReceiver =
+                bundle.getInputReceivers().get(timerId);
+            Preconditions.checkNotNull(fnTimerReceiver, "No FnDataReceiver found for %s", timerId);
+            try {
+              fnTimerReceiver.accept(timerValue);
+            } catch (Exception e) {
+              throw new RuntimeException(
+                  String.format(Locale.ENGLISH, "Failed to process timer: %s", timerValue));
+            }
+          });
+    }
   }
 
-  private void processElements(
-      Iterable<WindowedValue<InputT>> iterable, Collector<RawUnionValue> collector)
+  private void processElements(Iterable<WindowedValue<InputT>> iterable, RemoteBundle bundle)
       throws Exception {
-    checkState(
-        runtimeContext == getRuntimeContext(),
-        "RuntimeContext changed from under us. State handler invalid.");
-    checkState(
-        stageBundleFactory != null, "%s not yet prepared", StageBundleFactory.class.getName());
-    checkState(
-        stateRequestHandler != null, "%s not yet prepared", StateRequestHandler.class.getName());
+    Preconditions.checkArgument(bundle != null, "RemoteBundle must not be null");
+
+    String inputPCollectionId = executableStage.getInputPCollection().getId();
+    FnDataReceiver<WindowedValue<?>> mainReceiver =
+        Preconditions.checkNotNull(
+            bundle.getInputReceivers().get(inputPCollectionId),
+            "Main input receiver for %s could not be initialized",
+            inputPCollectionId);
+    for (WindowedValue<InputT> input : iterable) {
+      mainReceiver.accept(input);
+    }
+  }
 
-    try (RemoteBundle bundle =
-        stageBundleFactory.getBundle(
-            new ReceiverFactory(collector, outputMap), stateRequestHandler, progressHandler)) {
-      // TODO(BEAM-4681): Add support to Flink to support portable timers.
-      FnDataReceiver<WindowedValue<?>> receiver =
-          Iterables.getOnlyElement(bundle.getInputReceivers().values());
-      for (WindowedValue<InputT> input : iterable) {
-        receiver.accept(input);
+  private void fireEligibleTimers(
```

Comment:
I find that `fireEligibleTimers` says exactly that: Fire all timers which are ready to be fired.

It does not have the semantics of "fire all timers". It only happens to be the case because, beforehand, we advance processing and event time to the max timestamp.

I'll add a comment.